packages/container/lib/container.js

```diff
@@ -127,9 +135,86 @@ Container.prototype = {
    */
   lookupFactory(fullName, options) {
     assert('fullName must be a proper full name', this.registry.validateFullName(fullName));
+
+    if (isFeatureEnabled('container-factoryFor')) {
+      deprecate('Using "_lookupFactory" is deprecated. Please use container.factoryFor instead.', false, { id: 'container-lookupFactory', until: '2.12.0', url: 'TODO' });
+    }
+
     return factoryFor(this, this.registry.normalize(fullName), options);
   },
 
+  factoryFor(fullName, options = {}) {
+    let container = this;
+    let normalizedName = this.registry.normalize(fullName);
+    assert('fullName must be a proper full name', this.registry.validateFullName(normalizedName));
+
+    if (options.source) {
+      normalizedName = this.registry.expandLocalLookup(fullName, options);
+      // if expandLocalLookup returns falsey, we do not support local lookup
+      if (!normalizedName) { return; }
+    }
+
+    let factory = this.registry.resolve(normalizedName);
+
+    if (factory === undefined) { return; }
+
+    let manager = {
+      class: factory,
+      create(options = {}) {
+        let injections = injectionsFor(container, normalizedName);
+        let props = assign({}, injections, options);
+
+        runInDebug(() => {
+          let lazyInjections;
+          let validationCache = container.validationCache;
+          // Ensure that all lazy injections are valid at instantiation time
+          if (!validationCache[fullName] && factory && typeof factory._lazyInjections === 'function') {
+            lazyInjections = factory._lazyInjections();
+            lazyInjections = container.registry.normalizeInjectionsHash(lazyInjections);
+
+            container.registry.validateInjections(lazyInjections);
+          }
+
+          validationCache[fullName] = true;
+        });
+
+        this.class._toString = container.registry.makeToString(factory, fullName);
+
+        if (!this.class.create) {
+          throw new Error(`Failed to create an instance of '${normalizedName}'. Most likely an improperly defined class or` +
+                      ` an invalid module export.`);
+        }
+
+        if (this.class.prototype) {
+          injectDeprecatedContainer(this.class.prototype, container);
+        }
+
+
+        return this.class.create(props);
+      }
+    };
+
+    runInDebug(() => {
+      if (HAS_PROXY) {
+        let validator = {
+          get(obj, prop) {
+            if (prop !== 'class' && prop !== 'create') {
+              throw new Error(`You attempted to access "${prop}" on a factory manager created by container#factoryFor. "${prop}" is not a member of a factory manager."`);
+            }
+
+            return obj[prop];
+          },
+          set(obj, prop, value) {
+            throw new Error(`You attempted to set "${prop}" on a factory manager created by container#factoryFor. A factory manager is a read-only construct.`);
+          }
+        };
+        manager = new Proxy(manager, validator);
```

Comment:
Would like to have tests included that:

* getting a property that is not `class` or `create` throws an error
* setting a property throws an error
* calling `instanceof` with the return value from `factoryFor` throws an error

Per: https://github.com/emberjs/rfcs/blob/master/text/0000-factoryFor.md#development-mode-proxy